src/twisted/internet/test/test_unix.py

```diff
@@ -429,18 +444,91 @@ def _dataReceived(self, data):
         self.assertFalse(set(fdsToSend).intersection(set(proto.fds)))
+
+
+    def test_multiFileDescriptorReceivedPerRecvmsgOneCMSG(self):
+        """
+        _SendmsgMixin handles multiple file descriptors per recvmsg, calling
+        L{IFileDescriptorReceiver.fileDescriptorReceived} once per received
+        file descriptor. Scenario: single CMSG with two FDs.
+        """
+        from twisted.python.sendmsg import SCM_RIGHTS
+
+        def ancillaryPacker(fdsToSend):
+            ancillary = [(SOL_SOCKET, SCM_RIGHTS, pack('ii', *fdsToSend))]
+            expectedCount = 2
+            return ancillary, expectedCount
+
+        self._sendmsgMixinFileDescriptorReceivedDriver(ancillaryPacker)
+    if sendmsgSkip is not None:
+        test_multiFileDescriptorReceivedPerRecvmsgOneCMSG.skip = sendmsgSkip
+
+
+    def test_multiFileDescriptorReceivedPerRecvmsgTwoCMSGs(self):
+        """
+        _SendmsgMixin handles multiple file descriptors per recvmsg, calling
+        L{IFileDescriptorReceiver.fileDescriptorReceived} once per received
+        file descriptor. Scenario: two CMSGs with one FD each.
+        """
+        from twisted.python.sendmsg import SCM_RIGHTS
+
+        def ancillaryPacker(fdsToSend):
+            ancillary = [
+                (SOL_SOCKET, SCM_RIGHTS, pack('i', fd))
+                for fd in fdsToSend
+            ]
+            expectedCount = 2
+            return ancillary, expectedCount
+
+        self._sendmsgMixinFileDescriptorReceivedDriver(ancillaryPacker)
+    if platform.isMacOSX():
+        test_multiFileDescriptorReceivedPerRecvmsgTwoCMSGs.skip = (
+            "Multi control message ancillary sendmsg not supported on Mac.")
+    elif sendmsgSkip is not None:
+        test_multiFileDescriptorReceivedPerRecvmsgTwoCMSGs.skip = sendmsgSkip
+
+
+    def test_multiFileDescriptorReceivedPerRecvmsgBadCMSG(self):
+        """
+        _SendmsgMixin handles multiple file descriptors per recvmsg, calling
+        L{IFileDescriptorReceiver.fileDescriptorReceived} once per received
+        file descriptor. Scenario: unsupported CMSGs with no FDs.
+        """
+        # Given that we can't just send random/invalid ancillary data via the
+        # packer for it to be sent via sendmsg -- the kernel would not accept
+        # it -- we'll temporarily replace recvmsg with a fake one that produces
+        # a non-supported ancillary message level/type. This being said, from
+        # the perspective of the ancillaryPacker, all that is required it to
+        # let the test driver know that 0 file descriptors are expected.
+        from twisted.python import sendmsg
+
+        def ancillaryPacker(fdsToSend):
+            ancillary = []
+            expectedCount = 0
+            return ancillary, expectedCount
+
+        def fakeRecvmsgUnsupportedAncillary(skt, *args, **kwargs):
+            data = b'some data'
+            ancillary = [(SOL_SOCKET+1, sendmsg.SCM_RIGHTS+1, b'')]
```

Comment:
I faced this exact challenge when coding this - which values to use such that it consistently represents an unsupported level/type accross different platforms and Python versions.

Given that this code is exercising the not-supported handling, the first thing I did was improving the test such that it verifies the expected message was logged.

The other thing that can be done, since we're working at the Python level like you point out, is using non-`int` values in `fakeRecvmsgUnsupportedAncillary`: if we set `ancillary = [(None, None, b'')]` it will certainly trigger the not-supported case and be even better than an out-of-bounds-for-`int` which may become within-bounds-for-`int` in some future.

What do you say?